src/ciadmin/generate/branches.py

```diff
@@ -0,0 +1,72 @@
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+
+import os
+from asyncio import Lock
+
+import aiohttp
+from aiohttp_retry import ExponentialRetry, RetryClient
+from tcadmin.util.sessions import aiohttp_session
+
+_cache = {}
+_lock = {}
+
+
+# TODO: support private repositories. this will most likely require querying
+# GitHub as an app.
+async def get(repo_path, repo_type="git"):
+    """Get the list of branches present in `repo_path`. Only supported for
+    public GitHub repositories."""
+    if repo_type != "git":
+        raise Exception("Branches can only be fetched for git repositories!")
+
+    if repo_path.endswith("/"):
+        repo_path = repo_path[:-1]
+    branches_url = f"https://api.github.com/repos/{repo_path}/branches"
+
+    # 100 is the maximum allowed
+    # https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28#list-branches
+    params = {"per_page": 100}
+    headers = {}
+    if "GITHUB_TOKEN" in os.environ:
+        github_token = os.environ["GITHUB_TOKEN"]
+        headers["Authorization"] = f"Bearer {github_token}"
```

Comment:
Which token are you planning to use here, one of the automation user accounts?

Instead, what if we authenticated using a `taskcluster-firefoxci` app installation token? This way we don't need to add to the maintenance burden around Github user accounts. This is also guaranteed to be installed on any repo that we'd need to generate branches for, plus it has higher rate limits than a regular user token.

You can use https://github.com/mozilla-releng/simple-github to authenticate easily with an app installation (you'll  need to go to https://github.com/organizations/mozilla-releng/settings/apps/firefoxci-taskcluster to generate a private key.. I can do this for you if you don't have permission).

Reply:
I think it might be nice to use `simple-github` even if we don't use an app installation.. There's not currently a benefit to this other than simplifying code, but once https://github.com/mozilla-releng/simple-github/issues/20 is fixed we'll get some additional benefit.

Reply:
The main reason I was avoiding using an app installation was because it would make local runs prohibitively difficult, or we'd need to maintain a separate (token based) code path for them (unless we're willing to make people have their own local runs, or perhaps create a shared app for everyone to use). If there's a way to avoid this problem I'm all for using an app! It certainly has the added benefit of unlocking support for private repos here.

I agree that simple-github should be used regardless of whether we use an app or not. I didn't realize it supported non-app authentication, so I didn't consider it originally.

Reply:
I think supporting both would be pretty simple, the token method could be a fallback.

If we don't do the app installation, how does this work? Would we need to grant the `moz-automation` Github user access to every repository in Taskcluster? Or was there something easier you had in mind?

Reply:
On using an app, after your comments and some reflection I agree that it's probably no big deal to use an app in GHA, and a token for local use.

On private repositories, we simply do not support hook generation for them at the moment (see https://github.com/mozilla-releng/fxci-config/blob/6816b5c1ce88f1f32cc506e7114230fc622a307c/src/ciadmin/generate/in_tree_actions.py#L53-L56). Using an app in GHA would open up that possibility, but we would still run into issues in local usage unless everyone is using a token (or app) that has access to all configured private repositories. I'm honestly not sure what is better: don't support private repos at all (until we're forced to) or only support them in GHA. The latter runs the risk of oversights or bustage because we can't easily run diffs that are equivalent to what would be deployed.

Reply:
Oh, while I was rebasing this I realized the other reason that tokens are a bit easier: they're available automatically in GHA workflows. There's no secret creation or management to deal with. (Obviously the ones that GHA provides won't have access to private repos though.)

Reply:
Thanks, this sounds fine for now then! Were you still planning to switch to simple-github? We can leave that for a follow-up if you need to land this soon.

Reply:
I was going to, but I ended up discovering that I need access to headers for pagination, which [simple-github doesn't give me](https://github.com/mozilla-releng/simple-github/blob/2afb3503e75d789392aae7186d641d96cf7cdc17/src/simple_github/client.py#L142), so it's a bit of pain to use at the moment. Because of that, I'll leave it for a future change.